diffkemp/simpll/passes/SeparateCallsToBitcastPass.cpp

```diff
@@ -0,0 +1,138 @@
+        for (auto &Inst : BB) {
+            if (auto Call = dyn_cast<CallInst>(&Inst)) {
+                // Find call instructions calling a bitcast
+                // operator that directly corresponds to a function.
+                auto BitCast =
+                        dyn_cast<BitCastOperator>(Call->getCalledValue());
+
+                if (BitCast && isa<Function>(BitCast->getOperand(0))) {
+                    // Get the bitcasted function.
+                    auto srcFun =
+                            dyn_cast<Function>(BitCast->stripPointerCasts());
+
+                    // Ignore the instruction if the number of
+                    // arguments is lower than the number of parameters.
+                    if (Call->getNumArgOperands()
+                        < srcFun->getFunctionType()->getNumParams()) {
+                        continue;
+                    }
+
+                    // Ignore the instruction if the source function returns
+                    // void and the return value is used.
+                    if (srcFun->getReturnType()->isVoidTy()
+                        && Call->getType()->isVoidTy() && !Call->use_empty()) {
+                        continue;
+                    }
+
+                    // Bitcast all arguments to the types of source function
+                    // parameters. If the number of arguments is higher, use the
+                    // remaining arguments without bitcasting as a vargars is
+                    // present in the call instruction.
+                    std::vector<Value *> newArgs;
+                    auto paramType = srcFun->getFunctionType()->param_begin();
+                    auto endParamType = srcFun->getFunctionType()->param_end();
+                    auto arg = Call->arg_begin();
+                    auto endArg = Call->arg_end();
+
+                    while (paramType != endParamType) {
+                        if ((*arg)->getType() == *paramType) {
+                            newArgs.push_back(*arg);
+                        } else {
+                            // Bitcast the argument so that the types match.
+                            auto newArg = CastInst::Create(
+                                    Instruction::BitCast,
+                                    *arg,
+                                    *paramType,
+                                    "",
+                                    newArgs.empty() ? cast<Instruction>(Call)
+                                                    : cast<Instruction>(
+                                                            newArgs.back()));
+
+                            newArg->setDebugLoc(Call->getDebugLoc());
+                            newArgs.push_back(newArg);
+                        }
+
+                        ++paramType;
+                        ++arg;
+                    }
+
+                    // Add the remaining arguments if there are any.
+                    while (arg != endArg) {
```

Comment:
It doesn't make sense to have `endArg` as a separate variable if it is used only once. Use `Call->arg_end()` instead.